src/async_impl/client.rs

```diff
@@ -76,8 +76,10 @@ pub struct Client {
 
 /// A `ClientBuilder` can be used to create a `Client` with custom configuration.
 #[must_use]
-pub struct ClientBuilder {
+pub struct ClientBuilder<ConnectorLayers> {
```

Comment:
For my understanding, re: breaking change, is that so? I have it instantiated via `ClientBuilder::new()` or `ClientBuilder::default()` so that the the generic bounds can always be elided without caller specifying. I thought this was validated via my tests that don't specify anything.

That said it seems like it clean up the internal API / remove the external exposure to take the route you suggest, so that's a good reason regardless. Will look at this.

Reply:
Yea, anywhere that declares a value with the type would break, such as `fn extra_config(builder: ClientBuilder) -> ClientBuilder {}`.

Reply:
Can you make it defaulted
```rust
pub struct ClientBuilder<ConnectorLayers=Identity> {}
```?

This will fix backwards compatibility

Reply:
It would, but that would prevent accepting any builder but the original. I think for reqwest it makes sense to hide the generics, just keep an enum or box internally.

Reply:
An enum won't work, unfortunately, since it still would need to specify bounds if any variant is generic.

With boxing, I think there is a way to get it down to just a single layer of indirection, but it's awkward. The simpler ways to composite multiple layers would involve boxing at each layer. 

You could do a sub-builder that is generic or require layers be pre-composed, but that API doesn't feel great. Also it would be awkward to enforce it is only called once at compile time, because that usually would require generics too :) 

Another bad option is to return a wrapper build struct that does include the generics, when `connector_layer()` is called, and delegate a lot of calls to the inner. But that seems hideous to maintain if there will be more tower stacks yet, and also not a great experience for referencing in functions.

Reply:
Wondering about whether there are ways to move the type outside of the builder directly, and reference it... something like a static struct that we initialize and update, then consume inside `ClientBuilder` build call.

Reply:
I think there might be a way to do it with a generic const that holds the layer stack, which the build method then looks up. Awfully janky though :P

Reply:
I might be getting that wrong about boxing... theoretically all the layers are being consumed by the service, anyway, right, via the `layer()` call? So that vtable lookup is paid for, only once, then, but then we're back to known (but un-nameable) types in our `build()`, and no impact to the futures? Anyway we drop the boxed layers.

Reply:
It seems like trying to handle heterogenous layer stacks without generics is a pretty ugly proposition. You can type erase them every layer, at increasing runtime cost, or you can require that they be pre-flattened before input, which is an awkward API. 

After thinking about it more, I do see the appeal of Ariel's suggestion: of `pub struct ClientBuilder<ConnectorLayers=Identity> {}`. Granted it still does surface the types that we'd prefer not to leak.

It's true that this locks down naming the `ClientBuilder` type to only the no-layer case with the default, unless the caller starts specifying bounds. But it's not intrusive to the general userbase that doesn't care about custom layers, which seems most important?

And then for people who do want to use custom layers as well as name the `ClientBuilder` type, it's not much of a papercut to add a custom layer for the first time since the compiler will just suggest that they add the bounds. And we're into pretty niche usage at this point regardless.

The migration path to adding future tower stacks to other portions of reqwest seems fine too... just add another generic and also initialize another default of Identity.

Anyway, just sharing my initial thoughts before tuning out for the weekend. 

I certainly can do the type-erase-at-every-layer thing if we truly don't care about runtime impact.  I'll also keep poking at whether there is a way to get it down to one extra indirection layer, without forcing the layers to be pre-flattened by the caller. Maybe Ariel has some suggestions.

Reply:
Anyway, just sharing my initial thoughts before tuning out for the weekend.

I certainly can do the type-erase-at-every-layer thing if we prefer that regardless of the scaling runtime cost. We can always document that you can reduce allocations by passing in a flattened layer.

I'll also keep poking at whether there is a way to get it down to one extra indirection layer, without forcing the layers to be pre-flattened by the caller. Maybe Ariel has some suggestions.

Thanks for your time reviewing!

Reply:
If a user wanted to avoid boxing at every layer, they could pass in a stack of layers in one call, right? They would do the generic layering outside reqwest.

Reply:
Anyway I do see the appeal of Ariel's suggested `pub struct ClientBuilder<ConnectorLayers=Identity> {}` since it effectively only impacts callers that need to name the Client builder type *as well as* use custom layers. And at that point, the compiler will just tell them to add the bounds.

But, given that reqwest is even more general purpose than other libraries, and this is (hopefully) not that hot a path, probably just taking the runtime hit is the right call.

We can always document that you can avoid the extra indirection by pre-composing your stack before passing it in.

Reply:
>  If a user wanted to avoid boxing at every layer, they could pass in a stack of layers in one call, right? They would do the generic layering outside reqwest.

Ah yeah, you caught me stealth editing. This is true. We can certainly recommend it in the rust docs / snippets if you don't think that is too much of a complexity dump.

Reply:
Addressed this in a new commit, see top-level comment for details.